editoast/src/views/temporary_speed_limits.rs

```diff
@@ -0,0 +1,216 @@
+use axum::extract::Json;
+use axum::extract::State;
+use axum::Extension;
+use chrono::NaiveDateTime;
+use derivative::Derivative;
+use editoast_schemas::infra::DirectionalTrackRange;
+use serde::de::Error as SerdeError;
+use serde::{Deserialize, Serialize};
+use std::result::Result as StdResult;
+use utoipa::ToSchema;
+
+use crate::error::Result;
+use crate::models::temporary_speed_limits::TemporarySpeedLimit;
+use crate::models::Changeset;
+use crate::views::AuthorizationError;
+use crate::views::AuthorizerExt;
+use crate::CreateBatch;
+use crate::Model;
+use editoast_authz::BuiltinRole;
+
+use crate::AppState;
+
+crate::routes! {
+    "/temporary_speed_limits" => create,
+}
+
+editoast_common::schemas! {
+    TemporarySpeedLimitCreateForm,
+    TemporarySpeedLimitCreateResponse,
+    TemporarySpeedLimitItemForm,
+}
+
+#[derive(Serialize, Derivative, ToSchema)]
+struct TemporarySpeedLimitItemForm {
+    start_date_time: NaiveDateTime,
+    end_date_time: NaiveDateTime,
+    track_ranges: Vec<DirectionalTrackRange>,
+    speed_limit: i64,
+    obj_id: String,
+}
+
+#[derive(Serialize, Deserialize, ToSchema)]
+struct TemporarySpeedLimitCreateForm {
+    speed_limits: Vec<TemporarySpeedLimitItemForm>,
+}
+
+#[derive(Serialize, Deserialize, ToSchema)]
+struct TemporarySpeedLimitCreateResponse {
+    pub created_speed_limits_ids: Vec<i64>,
+}
+
+impl TemporarySpeedLimitItemForm {
+    pub fn into_temporary_speed_limit_changeset(self) -> Changeset<TemporarySpeedLimit> {
+        TemporarySpeedLimit::changeset()
+            .start_date_time(self.start_date_time)
+            .end_date_time(self.end_date_time)
+            .track_ranges(self.track_ranges)
+            .speed_limit(self.speed_limit)
+            .obj_id(self.obj_id)
+    }
+}
+
+impl<'de> Deserialize<'de> for TemporarySpeedLimitItemForm {
```

Comment:
That's unfortunate that we have to manually implement `Deserialize`, adding that much more code, only to check `end_date_time > start_date_time`. I found a way (not tested) to reduce the quantity of code with [this trick](https://stackoverflow.com/a/70293354) (which I honestly don't completely understand). I don't know what you think about it? In any case, feel free to resolve either way.

```rust
#[derive(Serialize, Deserialize, ToSchema)]
#[serde(remove = "Self")]
struct TemporarySpeedLimitItemForm {
    start_date_time: NaiveDateTime,
    end_date_time: NaiveDateTime,
    track_ranges: Vec<DirectionalTrackRange>,
    speed_limit: i64,
    obj_id: String,
}

impl<'de> Deserialize<'de> for TemporarySpeedLimitItemForm {
    fn deserialize<D>(deserializer: D) -> StdResult<TemporarySpeedLimitItemForm, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let this = Self::deserialize(deserializer)?;
        // Check dates
        if this.end_date_time > this.start_date_time {
            Ok(this)
        } else {
            Err(SerdeError::custom(format!(
                "The temporary_speed_limit start date '{}' must be before the end date '{}'",
                this.start_date_time, this.end_date_time
            )))
        }
    }
}
```

Reply:
I would be really stumped if this trick does work... (My first instinct is that rustc will complain about conflicting impls of Deserialize)

I also think we should write some guidelines about how to perform input data validation in editoast...

Reply:
I think I'll leave it as is until there is a consensus on the way we parse / validate data.